### PR TITLE
Tests UNKNOWN return status of conditional mutations

### PR DESCRIPTION
**Pull request by keith-turner**
These changes make it possible to instrument the manager and tablet server in integration tests to use TestAmple configured to randomly return UNKNOWN status for conditional mutations.

Servers were instrumented by modifying the AbstractServer constructor to accept a server context factory.  This allowed test version of Manager and Tsever to pass a factory that creates a server context that returns TestAmple.

A version of Comprehensive IT was updated to use the new instrumented servers.  This should provide some base coverage of UNKNOWN handling.  Still need to determine what Ample submit() calls are not being excercised and modify more ITs to use these instrumented servers.

### CODE REVIEW

File: test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.List;
+
+import org.apache.accumulo.core.client.ConditionalWriter;
+import org.apache.accumulo.core.data.ConditionalMutation;
+import org.apache.accumulo.test.ample.metadata.ConditionalWriterInterceptor;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlakyInterceptor implements ConditionalWriterInterceptor {
```
Comment (cshannon) [inline test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: This interceptor is a neat way to test the different possibilities. 

The current code is fine so this could be a follow on or just done if we actually need it in the future, I am wondering if we should we make the probabilities configurable? 

I was thinking it could be useful to add another constructor with config parameters to define what percentage of time to pick each action. It could be useful if you want to more heavily test something, or even if you just wanted the interceptor to only perform one action 100% of the time. 

In some cases having things be more deterministic could be useful but that might also be better written as another interceptor if we want deterministic testing since the goal of the Flaky testing is to be random outcomes.

**Review by cshannon — APPROVED** [review]
LGTM, I took a look at this in detail and overall some really nice changes here and enhancements for testing with TestAmple. 

When I first created TestAmple I hadn't considered the possibility of actually using the TestAmple implementation with the real metadata table and a real Manager like the flaky tests are doing. I was designing it for writing to a different table to not interfere with the real instance running.  

But setting the tests up this way, so that the real instance uses TestAmple and loads the flaky manager and interceptor, results in a really cool way to test that the the unknown status checks work generically across the board.

File: test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java (hunk 1 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: More determinism would be nice.  Seems like it would lead to more test, which is probably ok.  Like would have ComprehensiveFlakyAmple1IT that runs w/ pattern A, ComprehensiveFlakyAmple2IT that runs w/ pattern B for the FlakyInterceptor, etc.   Seems like some memory would be needed as multiple rows may be written w/ different collumns, so need reset/restart the pattern when that changes.

If a test fails the randomness may make it hard to debug.  It would probably be generally useful to add a bit more logging around handing of unknown as these should be rare, so it be nice if the code logged some breadcrumbs when they happen and it reacts to them.  That should be a quick change, will add a PR for that.

Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: Something else I am trying to figure out w/ this is coverage.  Want to know what code for handling unknown is not being exercised and add test to cover that.

Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: Yeah, the flaky test is nice to introduce randomness like a real system but it definitely makes it hard to know what broke and how to reproduce it. So anything that can be added to help track down the steps that caused the error would be helpful. 

There is one more issue with this that I just realized and that is the CI build is generally configured to re-run tests up to 5 times so handle flaky tests. So because of that, if we do get a test failure that is a rare edge case, it is unlikely Jenkins would catch it because it would just re-run and it would pass.

Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: > There is one more issue with this that I just realized and that is the CI build is generally configured to re-run tests up to 5 times to handle flaky tests. So because of that, if we do get a test failure that is a rare edge case, it is unlikely Jenkins would catch it because it would just re-run and it would pass.

Yeah that is a problem.  Should we open a follow on issue to explore adding more deterministic tests?

Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: This probably needs more discussion but I think the best approach might be to set up the surefire/failsure plugins differently. We really shouldn't need to run most tests with the `rerunFailingTestsCount` flag set as that just hides issues. I think a better approach would be to set up multiple profiles or executions for the test plugins and by default include all tests and don't set the  `rerunFailingTestsCount` flag. Then if we come across flaky tests, we could either try and fix them or if we can't we can `emporarily or permanently add them to another profile where we can set that flag (maybe we need our own -D property for the config). 

In general, the setting is certainly useful because sometimes tests are just hard to write and are flaky so it can be nice to not fail, but it seems like it should be a temporary solution until the test is fixed and not permanent. Keeping it on it definitely can be a problem as it will also hide true errors that need to be fixed but are intermittent.

@ctubbsii - Any thoughts on this and the best way we could set up our Maven test plugins with two different modes/profiles where we only execute a small list of tests with `rerunFailingTestsCount` flag if it's determined we need to? Or do you have another idea?

Comment (ctubbsii) [reply test/src/main/java/org/apache/accumulo/test/ample/FlakyInterceptor.java:35]: The two different modes that you're talking about are typically covered by a different test suite. Random behavioral tests have historically been covered by something like RandomWalk or ContinuousIngest, which moved to the accumulo-testing repository.

I don't think our integration tests should engage in behavior that results in a random outcome. If there's a random element in the test to cover an edge case, then the test itself should be written such that the edge case is guaranteed, or nearly guaranteed, to have coverage every time.

Example:

```java
@Test
public void randomTestBad() {
  coverEdgeCase20PercentOfTheTime();
}

@Test
public void randomTestBetter() {
  for (int x = 1; x < 100; x++) {
    coverEdgeCase20PercentOfTheTime();
  }
}
```

We can make the Maven build much more complex by having separate executions of different classes of tests, but I would recommend against doing that, because it can get way too complicated way too fast, and I think what we should be doing is writing more resilient ITs that have more comprehensive coverage of these edge cases.
